vkcs/internal/clients/config.go

```diff
@@ -0,0 +1,171 @@
+	maxRetriesCount         = 3
+	requestsMaxRetriesCount = 3
+	requestsRetryDelay      = 30 * time.Millisecond
+)
+
+// Config is interface to work with configer calls
+type Config interface {
+	LoadAndValidate() error
+	GetRegion() string
+	ComputeV2Client(region string) (*gophercloud.ServiceClient, error)
+	ImageV2Client(region string) (*gophercloud.ServiceClient, error)
+	NetworkingV2Client(region string, sdn string) (*gophercloud.ServiceClient, error)
+	PublicDNSV2Client(region string) (*gophercloud.ServiceClient, error)
+	BlockStorageV3Client(region string) (*gophercloud.ServiceClient, error)
+	KeyManagerV1Client(region string) (*gophercloud.ServiceClient, error)
+	ContainerInfraV1Client(region string) (*gophercloud.ServiceClient, error)
+	IdentityV3Client(region string) (*gophercloud.ServiceClient, error)
+	DatabaseV1Client(region string) (*gophercloud.ServiceClient, error)
+	SharedfilesystemV2Client(region string) (*gophercloud.ServiceClient, error)
+	LoadBalancerV2Client(region string) (*gophercloud.ServiceClient, error)
+	GetMutex() *mutexkv.MutexKV
+}
+
+// configer uses openstackbase.Config as the base/foundation of this provider's
+type configer struct {
+	auth.Config
+	ContainerInfraV1MicroVersion string
+}
+
+func ConfigureProvider(d *schema.ResourceData, terraformVersion string) (Config, diag.Diagnostics) {
+	config := &configer{
+		auth.Config{
+			Username:         d.Get("username").(string),
+			Password:         d.Get("password").(string),
+			TenantID:         d.Get("project_id").(string),
+			Region:           d.Get("region").(string),
+			IdentityEndpoint: d.Get("auth_url").(string),
+			UserDomainID:     d.Get("user_domain_id").(string),
+			UserDomainName:   d.Get("user_domain_name").(string),
+			AllowReauth:      true,
+			MaxRetries:       maxRetriesCount,
+			TerraformVersion: terraformVersion,
+			SDKVersion:       meta.SDKVersionString(),
+			MutexKV:          mutexkv.NewMutexKV(),
+		},
+		d.Get("cloud_containers_api_version").(string),
+	}
+
+	if config.UserDomainID != "" {
+		config.UserDomainName = ""
+	}
+
+	if err := config.LoadAndValidate(); err != nil {
+		return nil, diag.FromErr(err)
+	}
+	return config, nil
+}
+
+var _ Config = &configer{}
+
+// GetRegion is implementation of getRegion method
+func (c *configer) GetRegion() string {
+	return c.Region
+}
+
+func (c *configer) ComputeV2Client(region string) (*gophercloud.ServiceClient, error) {
+	return c.Config.ComputeV2Client(region)
+}
+
+func (c *configer) ImageV2Client(region string) (*gophercloud.ServiceClient, error) {
+	return c.Config.ImageV2Client(region)
+}
+
+func (c *configer) NetworkingV2Client(region string, sdn string) (*gophercloud.ServiceClient, error) {
+	client, err := c.Config.NetworkingV2Client(region)
+	if err != nil {
+		return client, err
+	}
+	if sdn != c.SearchInAllSDNs() {
+		client.MoreHeaders = map[string]string{
+			"X-SDN": sdn,
+		}
+	}
+	return client, err
+}
+
+func (c *configer) SearchInAllSDNs() string {
+	return "all"
```

Comment:
why not a public constant?

Reply:
From my point of view, providing user with function-based settings is more extensible way than relying on constants, but maybe constant is enough here.

Reply:
could we now remove this?